File: C++/Trapping Rain Water.cpp
class Solution {
public:
    int trap(int A[], int n) {
        int *leftmaxhigh = new int[n];
        int leftmax = 0;
        for(int i = 0; i < n; i++){
            if(A[i] > leftmax) leftmax = A[i];
            leftmaxhigh[i] = leftmax;
        }
        int rightmax = 0, sum = 0;
        for(int i = n - 1; i >= 0; i--){
            if(A[i] < rightmax) rightmax = A[i];
            if(min(rightmax, leftmaxhigh[i]) > A[i])
                sum += min(rightmax, leftmaxhigh[i]) - A[i];
        }
        return sum;
    }
};

File: Python/Trapping-Rain-Water.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def trap(self, A):
        leftmosthigh = [0 for i in range(len(A))]
        leftmax = 0
        for i in range(len(A)):
            if A[i] > leftmax: leftmax = A[i]
            leftmosthigh[i] = leftmax
        sum = 0
        rightmax = 0
        for i in reversed(range(len(A))):
            if A[i] > rightmax: rightmax = A[i]
            if min(rightmax, leftmosthigh[i]) > A[i]:
                sum += min(rightmax, leftmosthigh[i]) - A[i]
        return sum